janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java

```diff
@@ -1124,24 +1124,50 @@ public void executeMultiQuery(final Collection<InternalVertex> vertices, final S
         public Iterator<JanusGraphRelation> getNew(final VertexCentricQuery query) {
             InternalVertex vertex = query.getVertex();
             if (vertex.isNew() || vertex.hasAddedRelations()) {
-                return (Iterator) vertex.getAddedRelations(new Predicate<InternalRelation>() {
-                    //Need to filter out self-loops if query only asks for one direction
+                // Need to filter out self-loops if query only asks for one direction
 
-                    private JanusGraphRelation previous = null;
+                return new Iterator<JanusGraphRelation>() {
+                    Iterator<InternalRelation> iterator = vertex.getAddedRelations(t -> true).iterator();
+                    InternalRelation loop = null;
+                    InternalRelation current = null;
 
                     @Override
-                    public boolean apply(final InternalRelation relation) {
-                        if ((relation instanceof JanusGraphEdge) && relation.isLoop()
-                                && query.getDirection() != Direction.BOTH) {
-                            if (relation.equals(previous))
-                                return false;
+                    public boolean hasNext() {
+                        if (current == null) {
+                            current = nextRelation();
+                        }
+                        return current != null;
```

Comment:
(nitpick) we need to check for `current != null` only if it was `null` before calling this method. I.e. I would rewrite this method as:
```
@Override
public boolean hasNext() {
    if (current == null) {
        current = nextRelation();
        return current != null;
    }
    return true;
}
```